File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        Stack<Integer> stack = new Stack<>();
        int ans = 0;

        for(int i=0; i<height.length; i++){

            while(!stack.isEmpty() && height[i] > height[stack.peek()]){
                int middle = stack.pop();

                if(stack.isEmpty()) break;

                int left = stack.peek();
                int width = i - left - 1;

                int h = Math.min(height[left], height[i]) - height[middle];

                ans += width*h;
            }
            stack.push(i);
        }

        return ans;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        stack<int> stack;
        int ans = 0;

        for(int i=0; i<height.size(); i++){

            while(!stack.empty() && height[i] > height[stack.top()]){
                int middle = stack.top();
                stack.pop();

                if(stack.empty()) break;

                int left = stack.top();
                int width = i - left - 1;

                int h = min(height[i], height[left]) - height[middle];
                ans += width*h;
            }
            stack.push(i);
        }

        return ans;
    }
};